File: packages/provide-uterm-go/hub/ratelimit.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package hub

// MinRatePerSec is the tightest rate (tokens/sec) any bucket-backed policy may
// be configured with. Port of ratelimit.MIN_RATE_PER_SEC.
//
// One call per second is the tightest limit a [TokenBucket] can actually
// honour, because its burst is one second of its own rate. Below 1/sec the
// bucket's ceiling is under the single token a call costs, so it never holds a
// whole token and denies every call forever, however long the caller waits — a
// bricked endpoint dressed as a tight limit, and exactly what refusing a rate
// of 0 is meant to prevent. The server config therefore refuses any configured
// rate below this floor, and [RateLimiter] clamps to it: were a lower rate
// accepted, the clamp would silently hand back a *looser* limit than the
// operator wrote, and a rate limit that quietly loosens itself is worse than
// none.
const MinRatePerSec = 1.0

// TokenBucket is a simple token-bucket rate limiter. Port of
// provide.uterm.server.bridge.ratelimit.TokenBucket.
//
// It is not internally synchronised: the Python original relies on the single
// event loop, and this port relies on the composing [RateLimiter] holding its
// mutex across every Allow call. Standalone use must serialise Allow calls.
type TokenBucket struct {
	rate       float64
	burst      float64
	tokens     float64
	lastRefill float64
	clock      Clock
}

// NewTokenBucket builds a bucket refilling at ratePerSec tokens/sec. burst is
// the max burst size (nil defaults to ratePerSec, i.e. one second of burst).
// clock supplies the monotonic reading; nil selects the real clock.
func NewTokenBucket(ratePerSec float64, burst *float64, clock Clock) *TokenBucket {
	clock = orDefaultClock(clock)
	b := ratePerSec
	if burst != nil {
		b = *burst
	}
	return &TokenBucket{
		rate:       ratePerSec,
		burst:      b,
		tokens:     b,
		lastRefill: clock.Monotonic(),
		clock:      clock,
	}
}

// Allow consumes one token if available, returning true if the request is
// admitted. Tokens refill by elapsed*rate (capped at burst) since the last call.
func (b *TokenBucket) Allow() bool {
	now := b.clock.Monotonic()
	elapsed := now - b.lastRefill
	b.tokens = minFloat(b.burst, b.tokens+elapsed*b.rate)
	b.lastRefill = now
	if b.tokens >= 1.0 {
		b.tokens -= 1.0
		return true
	}
	return false
}

func minFloat(a, b float64) float64 {
	if a < b {
		return a
	}
	return b
}

func maxFloat(a, b float64) float64 {
	if a > b {
		return a
	}
	return b
}

File: packages/provide-uterm-server/src/provide/uterm/server/bridge/ratelimit.py
#
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
# SPDX-License-Identifier: AGPL-3.0-or-later
#
"""Token-bucket rate limiter for WebSocket message streams."""

from __future__ import annotations

import time

#: Tightest rate (tokens/sec) any bucket-backed policy may be configured with.
#:
#: This is 1.0 for a structural reason, not a taste one: :class:`TokenBucket`
#: defaults ``burst`` to one second of the rate, so a bucket configured below
#: 1.0 can never hold a whole token and therefore denies *every* call forever,
#: however long the caller waits. A rate in ``[0, 1)`` is a bricked endpoint
#: wearing the costume of a rate limit, so the server config refuses the whole
#: band rather than accepting a number that silently means "never".
#:
#: :class:`~provide.uterm.server.bridge.hub.limiter.RateLimiter` also clamps to
#: this floor. Config refusing below it keeps the clamp from quietly handing
#: back a *looser* limit than the operator wrote.
#:
#: Making sub-1 rates meaningful would mean decoupling burst from rate
#: (``burst = max(1.0, rate)``) — a change to token-bucket semantics across
#: every port and their recorded goldens. Worth doing deliberately if a
#: sub-1 policy is ever actually wanted; not worth doing by accident here.
MIN_RATE_PER_SEC = 1.0


class TokenBucket:
    """Simple token-bucket rate limiter.

    Args:
        rate_per_sec: Sustained token refill rate (tokens per second).
        burst: Maximum burst size.  Defaults to ``rate_per_sec`` (one second
            of burst capacity).
    """

    __slots__ = ("_burst", "_last_refill", "_rate", "_tokens")

    def __init__(self, rate_per_sec: float, burst: float | None = None) -> None:
        self._rate = float(rate_per_sec)
        self._burst = float(burst if burst is not None else rate_per_sec)
        self._tokens = self._burst
        self._last_refill = time.monotonic()

    def allow(self) -> bool:
        """Consume one token if available. Returns ``True`` if allowed."""
        now = time.monotonic()
        elapsed = now - self._last_refill
        self._tokens = min(self._burst, self._tokens + elapsed * self._rate)
        self._last_refill = now
        if self._tokens >= 1.0:
            self._tokens -= 1.0
            return True
        return False
